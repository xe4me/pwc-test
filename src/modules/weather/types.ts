/*
 * @interface Forecast
 * This interface represents a day's weather attributes
 * More information about each field can be found here https://www.weatherbit.io/api/swaggerui/weather-api-v2#!/Daily32Historical32Weather32Data/get_history_daily_city_city_country_country
 * */
export interface Forecast {
  wind_cdir: string;
  pres: number;
  high_temp: number;
  wind_gust_spd: number;
  wind_spd: number;
  dewpt: number;
  weather: {
    icon: string;
    code: number;
    description: string;
  };
  wind_dir: number;
  precip: number;
  low_temp: number;
  max_temp: number;
  datetime: string;
  temp: number;
  min_temp: number;
}
/*
 * @interface CurrentWeatherApiResponse
 * This interfaces is a minimal version what the API actually returns, all the non-required fields are omitted for brevity
 * */
export interface CurrentWeatherApiResponse {
  city_name: string;
  data: Forecast[];
}

/*
 * @enum Unit
 * Two types of units, Metric and Scientific
 * */
export enum Unit {
  Metric = 'm',
  Scientific = 's',
}
