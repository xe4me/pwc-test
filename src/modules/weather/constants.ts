export const DEFAULT_CITY = 'Sydney';
