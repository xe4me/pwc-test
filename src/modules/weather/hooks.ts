import useSWR from 'swr';
import { createEndpoint } from 'utils/createEndpoint';
import { fetcher } from 'utils/fetcher';
import { CurrentWeatherApiResponse, Unit } from './types';

export const useCurrentWeatherData = ({
  unit,
  city,
}: {
  unit: Unit;
  city: string;
}) => {
  const { data, error } = useSWR<CurrentWeatherApiResponse | null>(
    createEndpoint(
      `/forecast/daily?city=${city}&country=australia&state=nsw&days=7&units=${unit}`
    ),
    fetcher,
    {}
  );
  return { data, error };
};
