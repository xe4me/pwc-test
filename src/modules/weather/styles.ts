import { styled } from '@mui/material/styles';

import MuiContainer from '@mui/material/Container';

export const Container = styled(MuiContainer)(({ theme }) => ({
  position: 'relative',
  minHeight: `${theme.pxToRem(600)}`,
}));

export const Progress = styled('div')(() => ({
  position: 'absolute',
  top: '50%',
  left: '50%',
  transform: 'translate(-50%,-50%)',
}));
