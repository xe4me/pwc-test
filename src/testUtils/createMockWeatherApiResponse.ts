export const createMockWeatherApiResponse = (city_name: string) => ({
    city_name,
    data: [
      {
        wind_cdir: 'W',
        pres: 1014,
        high_temp: 20.2,
        wind_gust_spd: 5.9,
        wind_spd: 3.1,
        dewpt: 5.9,
        weather: {
          icon: 'c02d',
          code: 801,
          description: 'Few clouds',
        },
        wind_dir: 266,
        precip: 0,
        low_temp: 9.3,
        max_temp: 20.2,
        datetime: '2022-06-25',
        temp: 14.9,
        min_temp: 9.9,
      },
    ],
  });
