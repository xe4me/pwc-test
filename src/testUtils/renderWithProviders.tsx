import { render } from '@testing-library/react';
import { ReactElement } from 'react';
import { ThemeProvider } from '@mui/material';
import theme from '../theme';

export const renderWithProviders = (element: ReactElement) =>
  render(<ThemeProvider theme={theme}>{element}</ThemeProvider>);
