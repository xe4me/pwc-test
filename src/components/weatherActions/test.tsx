import React from 'react';
import { screen } from '@testing-library/react';
import { renderWithProviders } from 'testUtils/renderWithProviders';
import { Unit } from 'modules/weather/types';
import { axe } from 'jest-axe';
import { WeatherActions } from './index';

const renderComponent = () => {
  const onCityChangeSpy = jest.fn();
  const onUnitChangeSpy = jest.fn();
  return renderWithProviders(
    <WeatherActions
      unit={Unit.Metric}
      city="Sydney"
      onCityChange={onCityChangeSpy}
      onUnitChange={onUnitChangeSpy}
    />
  );
};
describe('<WeatherActions>', () => {
  describe('Snapshot and accessibility', () => {
    it('should match with the snapshot', () => {
      const { baseElement } = renderComponent();
      expect(baseElement).toMatchSnapshot();
    });
    it('should pass accessibility tests', async () => {
      const html = renderComponent();
      expect(await axe(html.container)).toHaveNoViolations();
    });
  });
  describe('Visible Elements', () => {
    it('should have the page title as Adjust your search', () => {
      renderComponent();
      expect(screen.queryByText('Adjust your search')).toBeInTheDocument();
    });
    it('should render the weather switch component', () => {
      renderComponent();
      expect(screen.queryByTestId('weather-unit-switch')).toBeInTheDocument();
    });
    it('should render the weather city input', () => {
      renderComponent();
      expect(screen.queryByTestId('weather-city-input')).toBeInTheDocument();
    });
  });
});
