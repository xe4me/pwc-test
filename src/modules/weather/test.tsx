import React from 'react';
import { screen } from '@testing-library/react';
import { createMockWeatherApiResponse } from 'testUtils/createMockWeatherApiResponse';
import { renderWithProviders } from 'testUtils/renderWithProviders';
import { axe } from 'jest-axe';
import { Weather } from './index';
import * as useCurrentWeatherDataHook from './hooks';
import { CurrentWeatherApiResponse } from './types';

const mockWeatherHook = (
  data: CurrentWeatherApiResponse | null,
  error: any
) => {
  jest
    .spyOn(useCurrentWeatherDataHook, 'useCurrentWeatherData')
    .mockImplementation(() => ({
      data,
      error,
    }));
};

describe('<Weather>', () => {
  describe('Snapshot and accessibility', () => {
    it('should match with the snapshot', () => {
      const { baseElement } = renderWithProviders(<Weather />);
      expect(baseElement).toMatchSnapshot();
    });
    it('should pass accessibility tests', async () => {
      const html = renderWithProviders(<Weather />);
      expect(await axe(html.container)).toHaveNoViolations();
    });
  });
  describe('Data fetching behaviours', () => {
    describe('When api errors', () => {
      it('should render "failed to load"', () => {
        mockWeatherHook(null, 'Something went wrong');
        renderWithProviders(<Weather />);
        expect(screen.queryByText('failed to load')).toBeInTheDocument();
      });
    });
    describe('When api succeeds', () => {
      it('should NOT render "failed to load"', () => {
        mockWeatherHook(createMockWeatherApiResponse('Sydney'), null);
        renderWithProviders(<Weather />);
        expect(screen.queryByText('failed to load')).not.toBeInTheDocument();
      });
    });
  });
});
