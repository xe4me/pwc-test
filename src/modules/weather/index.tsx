import { useState } from 'react';
import Paper from '@mui/material/Paper';
import Box from '@mui/material/Box';
import Divider from '@mui/material/Divider';
import { WeatherWeek } from 'components/weatherWeek';
import { WeatherHeader } from 'components/weatherHeader';
import { Spacing } from 'constants/spacing';
import { WeatherActions } from 'components/weatherActions';
import CircularProgress from '@mui/material/CircularProgress';
import * as Styled from './styles';
import { Unit } from './types';
import { useCurrentWeatherData } from './hooks';
import { DEFAULT_CITY } from './constants';

export function Weather() {
  const [unit, setUnit] = useState<Unit>(Unit.Metric);
  const [city, setCity] = useState<string>(DEFAULT_CITY);
  const { data, error } = useCurrentWeatherData({ unit, city });
  if (error) return <div>failed to load</div>; // @TODO Milad Implement proper error handling
  return (
    <Styled.Container maxWidth="lg">
      <Paper elevation={3}>
        {!data && !error && (
          // @TODO Milad : Ideally, the progress and the error both can be taken out and made more generic
          <Styled.Progress>
            <CircularProgress
              role="progressbar"
              aria-label="Weather data is loading ..."
              color="secondary"
            />
          </Styled.Progress>
        )}
        {data && (
          <Box p={Spacing.sm}>
            <WeatherHeader forecast={data.data[0]} city={data.city_name} />
            <Box mt={Spacing.md} mb={Spacing.md}>
              <Divider />
            </Box>
            <Box mt={Spacing.md} mb={Spacing.md}>
              <WeatherWeek daysForecasts={data.data} />
            </Box>
            <Box mt={Spacing.md} mb={Spacing.md}>
              <Divider />
            </Box>
            <WeatherActions
              unit={unit}
              onUnitChange={setUnit}
              city={city}
              onCityChange={setCity}
            />
          </Box>
        )}
      </Paper>
    </Styled.Container>
  );
}
