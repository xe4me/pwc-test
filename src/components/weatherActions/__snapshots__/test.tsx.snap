// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<WeatherActions> Snapshot and accessibility should match with the snapshot 1`] = `
<body>
  <div>
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-mhc70k-MuiGrid-root"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
      >
        <h5
          class="MuiTypography-root MuiTypography-h5 css-fru0ub-MuiTypography-root"
        >
          Adjust your search
        </h5>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-6 css-1fyyp8j-MuiGrid-root"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-8wd0g5-MuiPaper-root"
          data-testid="weather-unit-switch"
        >
          <div
            class="MuiFormControl-root css-1nrlq1o-MuiFormControl-root"
          >
            <label
              class="MuiFormLabel-root MuiFormLabel-colorPrimary css-u4tvz2-MuiFormLabel-root"
              id="unit-radio-group"
            >
              Unit
            </label>
            <div
              aria-labelledby="unit-radio-group"
              class="MuiFormGroup-root MuiFormGroup-row css-qfz70r-MuiFormGroup-root"
              data-testid="unit-radio-group"
              role="radiogroup"
            >
              <label
                class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd css-j204z7-MuiFormControlLabel-root"
              >
                <span
                  class="MuiRadio-root MuiRadio-colorPrimary MuiButtonBase-root MuiRadio-root MuiRadio-colorPrimary PrivateSwitchBase-root css-vqmohf-MuiButtonBase-root-MuiRadio-root"
                >
                  <input
                    class="PrivateSwitchBase-input css-1m9pwf3"
                    name="unit-radio-group"
                    type="radio"
                    value="s"
                  />
                  <span
                    class="css-hyxlzm"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1hbvpl3-MuiSvgIcon-root"
                      data-testid="RadioButtonUncheckedIcon"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                      />
                    </svg>
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1hhw7if-MuiSvgIcon-root"
                      data-testid="RadioButtonCheckedIcon"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M8.465 8.465C9.37 7.56 10.62 7 12 7C14.76 7 17 9.24 17 12C17 13.38 16.44 14.63 15.535 15.535C14.63 16.44 13.38 17 12 17C9.24 17 7 14.76 7 12C7 10.62 7.56 9.37 8.465 8.465Z"
                      />
                    </svg>
                  </span>
                  <span
                    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                  />
                </span>
                <span
                  class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-ahj2mt-MuiTypography-root"
                >
                  Imperial
                </span>
              </label>
              <label
                class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd css-j204z7-MuiFormControlLabel-root"
              >
                <span
                  class="MuiRadio-root MuiRadio-colorPrimary MuiButtonBase-root MuiRadio-root MuiRadio-colorPrimary PrivateSwitchBase-root Mui-checked css-vqmohf-MuiButtonBase-root-MuiRadio-root"
                >
                  <input
                    checked=""
                    class="PrivateSwitchBase-input css-1m9pwf3"
                    name="unit-radio-group"
                    type="radio"
                    value="m"
                  />
                  <span
                    class="css-hyxlzm"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1hbvpl3-MuiSvgIcon-root"
                      data-testid="RadioButtonUncheckedIcon"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                      />
                    </svg>
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-11zohuh-MuiSvgIcon-root"
                      data-testid="RadioButtonCheckedIcon"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M8.465 8.465C9.37 7.56 10.62 7 12 7C14.76 7 17 9.24 17 12C17 13.38 16.44 14.63 15.535 15.535C14.63 16.44 13.38 17 12 17C9.24 17 7 14.76 7 12C7 10.62 7.56 9.37 8.465 8.465Z"
                      />
                    </svg>
                  </span>
                  <span
                    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                  />
                </span>
                <span
                  class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-ahj2mt-MuiTypography-root"
                >
                  Metric
                </span>
              </label>
            </div>
          </div>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-6 css-1fyyp8j-MuiGrid-root"
      >
        <form
          data-testid="weather-city-input"
        >
          <div
            class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-s8p2ph-MuiPaper-root"
          >
            <div
              class="MuiFormControl-root css-1nrlq1o-MuiFormControl-root"
            >
              <label
                class="MuiFormLabel-root MuiFormLabel-colorPrimary MuiFormLabel-filled css-u4tvz2-MuiFormLabel-root"
                id="weather-city-input"
              >
                City
              </label>
              <div
                aria-labelledby="weather-city-input"
                class="MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-1le28he-MuiInputBase-root"
              >
                <input
                  aria-invalid="false"
                  aria-label="City input"
                  class="MuiInputBase-input css-yz9k0d-MuiInputBase-input"
                  placeholder="Enter a city in NSW"
                  type="text"
                  value="Sydney"
                />
              </div>
            </div>
            <button
              aria-label="Search button"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-1gtpq4r-MuiButtonBase-root-MuiIconButton-root"
              tabindex="0"
              type="submit"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="SearchIcon"
                focusable="false"
                role="img"
                viewBox="0 0 24 24"
              >
                <path
                  d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </form>
      </div>
    </div>
  </div>
</body>
`;
